### PR TITLE
Delete the makefile build system

### PR DESCRIPTION
**Pull request by alexcrichton**
This PR deletes the makefile build system in favor of the rustbuild build system. The beta has now been branched so 1.16 will continue to be buildable from the makefiles, but going forward 1.17 will only be buildable with rustbuild.

Rustbuild has been the default build system [since 1.15.0](https://github.com/rust-lang/rust/pull/37817) and the makefiles were [proposed for deletion](https://internals.rust-lang.org/t/proposal-for-promoting-rustbuild-to-official-status/4368) at this time back in November of last year.

And now with the deletion of these makefiles we can start getting those sweet sweet improvements of using crates.io crates in the compiler!

### CODE REVIEW

**Comment by steveklabnik** [comment]
> And now with the deletion of these makefiles we can start getting those sweet sweet improvements of using crates.io crates in the compiler!

I know I already did a 🎊 reaction, but this deserves many more.

🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊 🎊

**Comment by petrochenkov** [comment]
Please fix https://github.com/rust-lang/rust/issues/36385 before removing the old build system.

I planned to do this myself, but I won't be able to do this until this PR is merged.

**Comment by brson** [comment]
The patch looks fine to me.

@petrochenkov #36385 seems like both a difficult problem to solve and one that isn't causing a great deal of pain (Rust devs are still developing and nobody has fixed it). Should we really stop and tackle that right now?

**Comment by petrochenkov** [comment]
@brson 
@alexcrichton already [suggested a solution](https://github.com/rust-lang/rust/issues/36385#issuecomment-266301521), is well familiar with the build and is a master of hacking up kinda working systems quickly.
My estimation for the simplest (but very helpful) solution with timestamps (one timestamp for a whole test group, aux files may be not considered at all) is a couple of hours.

>Rust devs are still developing

That's because the old build systems still exists! Otherwise, they probably don't develop on Windows, or irresponsibly submit untested PRs.

**Comment by alexcrichton** [comment]
Let's just put that issue to rest, a commit's now attached to this PR. I was hoping to give someone else a chance to get their feet wet with the build system and/or compiletest, but oh well.

**Comment by alexcrichton** [comment]
@petrochenkov FWIW I've found the way you've approached #36385 as somewhat frustrating and unproductive. You've repeatedly brought it up and as you've notice I tried to outline clear solutions. There's been ample time (2 whole release cycles) where the deprecation of the makefiles has been announced, so this PR shouldn't come as a surprise.

In the future it'd be great to take a more proactive and less passive-aggressive stance towards solving these sorts of issues.

**Comment by ishitatsuyuki** [comment]
There's many critical issues breaking shared linking build, like #39215.

Please make a list to track them.

**Comment by pnkfelix** [comment]
@alexcrichton 

> Let's just put that issue to rest, a commit's now attached to this PR. I was hoping to give someone else a chance to get their feet wet with the build system and/or compiletest, but oh well.

Wait, why isn't that [commit](https://github.com/rust-lang/rust/pull/39431/commits/39321aab4ea84d93b9fcce74e3870532187e39c8) a separate PR?

Surely that would be a step forward to let people concerned about the makefiles going away verify that they can use rustbuild before landing this whole PR too?

**Comment by alexcrichton** [comment]
@pnkfelix well I don't really mind any particular organization, I'd just like to land this!

**Comment by petrochenkov** [comment]
@pnkfelix 
I've cherry-picked https://github.com/rust-lang/rust/pull/39431/commits/39321aab4ea84d93b9fcce74e3870532187e39c8 and tested it locally ([report](https://github.com/rust-lang/rust/issues/36385#issuecomment-277131231)).

**Comment by ishitatsuyuki** [comment]
What are you guys doing?
Rust 1.15 came to stable this morning, you guys have been already annoyed package maintainers by releasing a \*\*\*\* flawed buildsystem.

Please, delay the removal until the (dynamic) linking issues are fixed.

**Comment by alexcrichton** [comment]
@ishitatsuyuki as [explained in the PR](https://github.com/rust-lang/rust/pull/39431#issue-204434112) this is a long time coming. We've still got 3 months before this hits stable anyway to fix more issues as they arise.

Rewriting a build system is almost guaranteed to not be bug-for-bug compatible, so many quirks and bugs of the makefiles are not carried over (either accidentally or intentionally). Issues like #39215 were barely ever supported by the makefiles (if at all, only by luck) and aren't really a great reason to completely block this in my opinion.

These bug fixes tend to just be a line here or there, and PRs are always welcome!

**Comment by ishitatsuyuki** [comment]
@alexcrichton I'm concerned about breaking distro packaging entirely by only testing on rustup.

A buildsystem definitely should not be flawed (flawed by *default* is annoying enough though), and we will completely lose dynamic linking if you removed the old one.

Be patient. You didn't tend to break old rust code, but you're going to break up-to-date distro build script now. (By the way, this would also break the rust-git package in AUR)

**Comment by ishitatsuyuki** [comment]
Issue unblocked, I'm okay with the merge.

**Comment by bstrie** [comment]
As much as I'd love to forge ahead with Cargo, I agree with @ishitatsuyuki that we should make explicit overtures to Rust packagers to make sure they have everything they need to exist in a world without makefiles.

File: src/etc/apple-darwin.supp [hunk 1]
```diff
@@ -1,75 +0,0 @@
-{
```
Comment (tamird) [inline src/etc/apple-darwin.supp]: how come these files are no longer needed? where do valgrind suppressions come from these days? I grepped around and couldn't find it.

**Comment by petrochenkov** [comment]
@alexcrichton 
The test caching commit has suddenly disappeared after rebase.

**Comment by ishitatsuyuki** [comment]
Please also make sure that you have translated recent changes like #39425 into rustbuild.

**Comment by alexcrichton** [comment]
@petrochenkov oops mixed up the branches

@ishitatsuyuki yes as you can see in the changes made in #39425, rustbuild is already changed.

@bors: r=brson

**Comment by tamird** [comment]
@alexcrichton can you help me understand why the valgrind suppressions are no longer needed?

**Comment by alexcrichton** [comment]
@tamird oh sorry I missed that. We haven't used those suppressions in ages and they're just ancient artifacts at this point. I'd be quite surprised if they're at all accurate/working any more

**Comment by tamird** [comment]
Oh, I remember now; these suppressions were used when running the entire
suite under valgrind. I bet nobody's done that since I last did >.<

4 февр. 2017 г. 14:35 пользователь "Alex Crichton" <notifications@github.com>
написал:

**Comment by frewsxcv** [comment]
This PR might have caused this issue during the rollup: https://travis-ci.org/rust-lang/rust/jobs/198487299

**Comment by frewsxcv** [comment]
It looks like running 'try' does *not* test all environments, so it's unlikely to reproduce the MIPS issue

**Comment by alexcrichton** [comment]
@bors: r=brson

FWIW I don't think `try` really works for us, although I have no idea why

**Comment by vi** [comment]
Does this (crate.io-ifying rustc) mean boostrapping Rust from ground up got harder or got easier?

**Comment by cuviper** [comment]
Before, you only needed cross-compiled `rustc` to natively bootstrap on a new platform.  Now you'll need `rust-std` and `cargo` too.

**Comment by vi** [comment]
@cuviper , "from the ground up" means without using anything that is Rust-related pre-built. For example, while also doing security audit.

File: src/etc/make-win-dist.py [hunk 2]
```diff
@@ -1,124 +0,0 @@
-# Copyright 2013-2014 The Rust Project Developers. See the COPYRIGHT
```
Comment (tamird) [inline src/etc/make-win-dist.py]: looks like this file is still needed

**Comment by aidanhs** [comment]
Is there an issue somewhere for tracking feature parity between rustbuild and the makefiles or shall I raise one? I was curious so had a quick look over the weekend and noticed that `--disable-stage0-landing-pads` doesn't work (along with some others).

**Comment by alexcrichton** [comment]
@bors: r=brson

@tamird alas well we'll get rid of that file one day!

@aidanhs feel free to file issues! At this point we're in 'issue per bug' territory, not 'pile on tracking issue'

**Comment by tamird** [comment]
@alexcrichton re-adding that file in the last commit means that its git history is obliterated. Perhaps you could avoid its removal altogether?

**Comment by cuviper** [comment]
@vi, so you'd have to rebuild the entire chain from before it was self-hosting, written in OCaml?  I guess the added challenge with rustbuild may be in locating historical crates.io packages, but note that those sources *are* vendored in the rust tree.  You might also need that vendoring for cargo though.  I suggest you start a thread on the internals forum if this is something you want to be able to do.

**Comment by vi** [comment]
@cuviper, As far as I know there are projects for making Rust bootstrappable without repeating it's entire history from Ocaml across all intermediate releases. Like a simplified non-verifying non-optimizing Rust interpreter or compiler written in C.

**Comment by frewsxcv** [comment]
@alexcrichton Does [this fail from my rollup](https://github.com/rust-lang/rust/pull/39609#issuecomment-277904822) seem relevant to your changes here?

**Comment by alexcrichton** [comment]
@frewsxcv nah that's #38878

**Comment by Nashenas88** [comment]
Shouldn't the main README.md also be updated? It still says to use `make`.

**Comment by alexcrichton** [comment]
Nah we still have a bare-bones makefile at least for familiarity, so that documentation is still correct. We may wish to feature `x.py` more prominently though!

**Comment by J-F-Liu** [comment]
It will also be good to split current compiler into several crates, e.g. lexer, parser, mir etc. Crates can compete with each other which results in a better compiler.

**Comment by steveklabnik** [comment]
@J-F-Liu this is already true, check the subdirectories of https://github.com/rust-lang/rust/tree/master/src, for example, https://github.com/rust-lang/rust/blob/master/src/librustc_mir/Cargo.toml is for MIR.

**Pull request merged**